app/src/main/java/de/westnordost/streetcomplete/measure/MeasureActivity.kt

```diff
@@ -0,0 +1,418 @@
+        suspendCancellableCoroutine { cont ->
+            val dlg = AlertDialog.Builder(this)
+                .setTitle(R.string.no_camera_permission_warning_title)
+                .setMessage(R.string.no_camera_permission_warning)
+                .setPositiveButton(android.R.string.ok) { _, _ -> cont.resume(true) }
+                .setNegativeButton(android.R.string.cancel) { _, _ -> cont.resume(false) }
+                .setOnCancelListener { cont.resume(false) }
+                .create()
+            cont.invokeOnCancellation { dlg.cancel() }
+            dlg.show()
+        }
+
+    /* ----------------------------------------- Intent ----------------------------------------- */
+
+    companion object {
+        private const val PARAM_MEASURE_MODE = "measure_mode"
+        private const val PARAM_DISPLAY_UNIT = "display_unit"
+        private const val PARAM_DISPLAY_UNIT_PRECISION = "display_unit_precision"
+        private const val PARAM_REQUEST_RESULT = "request_result"
+
+        private const val METER = "meter"
+        private const val FOOT_AND_INCH = "foot_inch"
+
+        /** The action to identify a result */
+        const val RESULT_ACTION = "de.westnordost.streetcomplete.measure.RESULT_ACTION"
+        /** The exact floating point measure result. It is up to the caller to round and/or convert
+         *  it according to his preference */
+        const val RESULT_MEASURE_IN_METERS = "measure_result"
+
+        /** Create the intent for starting this activity, with optional parameters:
+         * @param mode specifies whether to measure distances on the ground, on walls or both
+         * @param unit specifies which unit (meters or foot/inch) should be used for display and
+         *             with which precision the measure should be shown (where to round to).
+         * @param requestResult whether this activity should return a result. If yes, the activity
+         *                      will return the raw measure result (not rounded) in
+         *                      RESULT_MEASURE_IN_METERS when the user confirmed it. */
+        fun createIntent(
+            context: Context,
+            mode: MeasureMode? = null,
+            unit: MeasureDisplayUnit? = null,
+            requestResult: Boolean? = null
+        ): Intent {
+            val intent = Intent(context, MeasureActivity::class.java)
+            mode?.let { intent.putExtra(PARAM_MEASURE_MODE, it.name) }
+            when(unit) {
+                is MeasureDisplayUnitFeetInch -> {
+                    intent.putExtra(PARAM_DISPLAY_UNIT, FOOT_AND_INCH)
+                    intent.putExtra(PARAM_DISPLAY_UNIT_PRECISION, unit.inchStep)
+                }
+                is MeasureDisplayUnitMeter -> {
+                    intent.putExtra(PARAM_DISPLAY_UNIT, METER)
+                    intent.putExtra(PARAM_DISPLAY_UNIT_PRECISION, unit.decimals)
+                }
+                null -> {}
+            }
```

Comment:
Would it make sense to make the `MeasureDisplayUnit` interface serializable? Then its (de)serialization code would be defined there, where it's more relevant.

Reply:
Hmm, good idea